src/query/storages/common/stage/src/read/columnar/projection.rs

```diff
@@ -0,0 +1,124 @@
+            .fields()
+            .iter()
+            .position(|f| f.name() == field_name)
+        {
+            Some(pos) => {
+                pushdown_columns.push(pos);
+                let from_field = input_schema.field(pos);
+                let expr = Expr::ColumnRef {
+                    span: None,
+                    id: pos,
+                    data_type: from_field.data_type().into(),
+                    display_name: from_field.name().clone(),
+                };
+
+                // find a better way to do check cast
+                if from_field.data_type == to_field.data_type {
+                    expr
+                } else if can_cast_types(
+                    Field::from(from_field).data_type(),
+                    Field::from(to_field).data_type(),
+                ) {
+                    check_cast(
+                        None,
+                        false,
+                        expr,
+                        &to_field.data_type().into(),
+                        &BUILTIN_FUNCTIONS,
+                    )?
+                } else {
+                    return Err(ErrorCode::BadDataValueType(format!(
+                        "fail to load file {}: Cannot cast column {} from {:?} to {:?}",
+                        location,
+                        field_name,
+                        from_field.data_type(),
+                        to_field.data_type()
+                    )));
+                }
+            }
+            None => {
+                match null_as {
+                    // default
+                    NullAs::Error => {
+                        return Err(ErrorCode::BadDataValueType(format!(
+                            "file {} missing column `{}`",
+                            location, field_name,
+                        )));
+                    }
+                    NullAs::Null => {
+                        if to_field.is_nullable() {
+                            Expr::Constant {
+                                span: None,
+                                data_type: to_field.data_type().into(),
+                                scalar: Scalar::Null,
+                            }
+                        } else {
+                            return Err(ErrorCode::BadDataValueType(format!(
+                                "{} missing column `{}`",
+                                location, field_name,
+                            )));
+                        }
+                    }
+                    NullAs::FieldDefault => {
+                        let default_values = &default_values.as_deref().expect(
+                            "default_values should not be none when miss_field_as=FIELD_DEFAULT",
+                        );
+                        default_values[i].as_expr(&BUILTIN_FUNCTIONS)
+                    }
+                }
+            }
+        };
+        if !matches!(expr, Expr::Constant { .. }) {
+            num_inputs += 1;
```

Comment:
we can't check input by `Expr::Constant`, because default value may be a function. `num_inputs` should increase by one if the `pos` is some.

Reply:
Alternatively, we can use `pushdown_columns` instead of `num_inputs`. If `pushdown_columns` is empty, it means `num_inputs` is 0.

Reply:
fixed, thank you!